### ISSUE TITLE
Generalize Metric to any loss function

### ISSUE THREAD

**Issue by phcerdan** [Project-MONAI/MONAI#3694]
**Is your feature request related to a problem? Please describe.**
If I introduce a new loss function, I have to change the metric as well. For example: `monai.losses.Dice` and `monai.metrics.DiceMetric`. Most of the losses have not a metric associated.

**Describe the solution you'd like**
I wonder if the metrics can be made general to any loss function. Where the loss function would be an argument of the metric. Other metric arguments would be the reduction method and similar things.

**Describe alternatives you've considered**
The alternative is to develop my own Metric every time I change the loss function. But not ideal imo.

**Comment by umgpy** [comment]
Hi, I am facing a similar issue. For my case I would like to have two or more metrics (eg-Dice, HD95) along with dice as loss. 
However,if both the metrics are computed the loss after couple of epochs results in NaN.

Would love to find out how and if the metrics could be generalized with any loss as @phcerdan mentioned.

**Comment by umgpy** [comment]
Hi, I am facing a similar issue. For my case I would like to have two or more metrics (eg-Dice, HD95) along with dice as loss. 
However,if both the metrics are computed the loss after couple of epochs results in NaN.

Would love to find out how and if the metrics could be generalized with any loss as @phcerdan mentioned.

**Comment by wyli** [comment]
great suggestion, the main engineering value-added for monai metric is the feature of computing results from mini-batches and aggregate (potentially over multiple processes) these results at the epoch level (e.g. https://github.com/Project-MONAI/tutorials/blob/master/modules/compute_metric.py). I think you can use the `CumulativeAverage` to achieve averaging for the monai loss functions:
https://github.com/Project-MONAI/MONAI/blob/c2018cf620a7ddf2130a272bf8a1b68ced8f2e42/monai/metrics/cumulative_average.py#L30-L36
the cumulative results could be used for tensorboard plotting/early stopping. Or, is there any specific metric use case that difficult to achieve with the current API?

**Comment by umgpy** [comment]
@wyli Thanks for getting back!
More than early stopping, I would like to use multiple metrics for reporting. For example, use Dice loss along with DSC metric for validation and saving the best model but at the same time make use of one or more additional metric for reporting/ comparison and/or ranking  as specified in  [here](https://github.com/emrekavur/CHAOS-evaluation/blob/master/CHAOS_Metrics_and_Evaluation_Method.pdf)

Example, after each epoch,
Dice = 0.8635, HD95 = 12.2, ASSD = 8.3 (for reporting)

But on computing the extra metrics the loss goes to nan after a couple of epoch.

Normal- 
_epoch 2 average loss: 0.6035
saved new best metric model
current epoch: 2 current mean dice: 0.3898 BG: 0.6600 fem: 0.1438 pel: 0.3656
best mean dice: 0.3898 at epoch: 2
current epoch: 2 current mean hd95: 94.7535 BG: 54.7478 fem: 124.4526 pel: 105.0602
current epoch: 2 current mean assd: 18.3567 BG: 10.2701 fem: 25.0300 pel: 19.7702
time consuming of epoch 2 is: 1576.0210_


Intial error-
_
epoch 11 average loss: 0.5795

/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 0 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/numpy/lib/function_base.py:4486: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 1 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 2 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 0 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 1 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/umgpy/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 2 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")

current epoch: 11 current mean dice: 0.1033 BG: 0.1736 fem: 0.0642 pel: 0.0722
best mean dice: 0.4128 at epoch: 6
current epoch: 11 current mean hd95: 89.1498 BG: 39.9831 fem: 141.0331 pel: 86.4331
best mean dice: -1.0000 at epoch: -1
current epoch: 11 current mean assd: inf BG: inf fem: inf pel: inf
best mean dice: -1.0000 at epoch: -1
time consuming of epoch 11 is: 639.5590_


Finally,
_5/8, train_loss: nan, step time: 1.5168
6/8, train_loss: 0.5702, step time: 1.5253
7/8, train_loss: 0.5762, step time: 1.5282
8/8, train_loss: nan, step time: 1.5164
epoch 14 average loss: nan
current epoch: 14 current mean dice: 0.1034 BG: 0.1757 fem: 0.0648 pel: 0.0697
best mean dice: 0.4128 at epoch: 6
current epoch: 14 current mean hd95: 89.3620 BG: 40.0747 fem: 141.1014 pel: 86.9099
current epoch: 14 current mean assd: inf BG: inf fem: inf pel: inf
time consuming of epoch 14 is: 639.2755_

I am using `DiceMetric`, `HausdorffDistanceMetric` and `SurfaceDistanceMetric`. Using them individually is not an issue though.

With regard to using `CumulativeAverage`, since  I do not need all of them averaged it is not much use. Let me know if I am missing something. Would prefer to have additional metrics calculated in a single run than run it multiple times for each of them indivdually.
Thanks

**Comment by wyli** [comment]
> But on computing the extra metrics the loss goes to nan after a couple of epoch.

Perhaps you can detach the prediction to prevent backprop before computing alternative loss values?

**Comment by phcerdan** [comment]
> So the cumulative APIs ([avg](https://github.com/Project-MONAI/MONAI/blob/c2018cf620a7ddf2130a272bf8a1b68ced8f2e42/monai/metrics/cumulative_average.py#L20), [generic](https://github.com/Project-MONAI/MONAI/blob/c2018cf620a7ddf2130a272bf8a1b68ced8f2e42/monai/metrics/metric.py#L112-L115)) is the generic metric wrapper that you are looking for.

It is indeed, thanks @wyli.
I still fail to fully understand the purpose of `DiceMetric`. Is there any difference between `DiceMetric` and doing a `DiceLoss` + `CumulativeAverage`? Thanks!

**Comment by myron** [comment]
> **Is your feature request related to a problem? Please describe.** If I introduce a new loss function, I have to change the metric as well. For example: `monai.losses.Dice` and `monai.metrics.DiceMetric`. Most of the losses have not a metric associated.
> 
> **Describe the solution you'd like** I wonder if the metrics can be made general to any loss function. Where the loss function would be an argument of the metric. Other metric arguments would be the reduction method and similar things.
> 
> **Describe alternatives you've considered** The alternative is to develop my own Metric every time I change the loss function. But not ideal imo.

HI @phcerdan,  loss and metric are two different things, and it's usually not possible to define a generic correspondence.  Dice loss is used as an objective function for the Optimizer.  Dice loss value is computed as a soft dice , between probability maps and ground truth labels.  Where as Dice metric is a proper dice value, and computed after taking argmax() (to first convert soft probability maps to the class indices and then computing dice value).   

Generally, there are many loss functions you can use , e.g. cross-entropy loss, dice_focal loss, or your own loss. But in all those cases, you still care about a generic dice metric value (or some other metric).  You may have a combination (weighted sum of several losses), but you metric does not change (usually dice metric).

Loss function (such as Dice loss or cross-entropy loss) is an objective function for the optimizer, it's absolute value is usually not important  (aside from the fact that we want to minimize that value, and during the optimization , we expect it to decrease)

For an arbitrary loss function, there is no corresponding metric  function. 
Does this answer your question? thank you

**Comment by myron** [comment]
> may result in nan/inf distance.

@umgpy  it sounds like you are asking a completely different question in the same issue.  it sounds like you want to  report several metrics (Dice metric, Hausdorf.. metric, Surface.. metric) but select the best checkpoint  based on only Dice metric?   I do not see a problem here, just compute manually all metrics you want, and in your validation loop only use the dice value to save the checkpoint. 

If you're not writing your own training  / validation loops manually, what do you use? it seems your issue is related to some "trainer" class that you chose.
